pkg/tests/service/task_test.go

```diff
@@ -64,3 +64,64 @@ func TestTaskSchedulerCanAllocateTask(t *testing.T) {
 	}
 	t.Fatalf("task not allocated")
 }
+
+func TestTaskSchedulerCanReallocateTask(t *testing.T) {
+	taskStorage := taskservice.NewMemTaskStorage()
+	taskService := taskservice.NewTaskService(taskStorage, nil)
+
+	dnSvcNum := 1
+	cnSvcNum := 2
+	opt := DefaultOptions().
+		WithDNServiceNum(dnSvcNum).
+		WithCNServiceNum(cnSvcNum).
+		WithTaskStorage(taskStorage)
+
+	// initialize cluster
+	c, err := NewCluster(t, opt)
+	require.NoError(t, err)
+
+	// start the cluster
+	err = c.Start()
+	require.NoError(t, err)
+
+	err = taskService.Create(context.TODO(), task.TaskMetadata{ID: "a"})
+	require.NoError(t, err)
+	tasks, err := taskService.QueryTask(context.TODO(),
+		taskservice.WithTaskStatusCond(taskservice.EQ, task.TaskStatus_Created))
+	require.NoError(t, err)
+	require.Equal(t, 1, len(tasks))
+
+	var uuid1 string
+	for i := 0; i < 100; i++ {
+		t.Logf("iteration: %d", i)
+		tasks, err := taskService.QueryTask(context.TODO(),
+			taskservice.WithTaskStatusCond(taskservice.EQ, task.TaskStatus_Running))
+		require.NoError(t, err)
+
+		if len(tasks) == 1 {
+			t.Logf("task %d allocated on %s", tasks[0].ID, tasks[0].TaskRunner)
+			uuid1 = tasks[0].TaskRunner
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
```

Comment:
Use a loop to wait for the desired state of the task until it times out. In a CI environment, sleep may not be long enough and cause the test to fail

Reply:
The error is not raised from here. I believe it is from `pkg/tests/service/service.go#314`
```
	if c.opt.initial.cnServiceNum != 0 {
		time.Sleep(10 * time.Second) // this line
		if err := c.startCNServices(); err != nil {
			return err
		}
	}
```
It seems that 10 seconds is not enough for CN to get DN's shard information in CI environment. Then at `pkg/vm/engine/txn/shard.go#83`,
```
	for _, store := range clusterDetails.DNStores {
		info := store.Shards[0] // this line
		shards = append(shards, Shard{
			DNShardRecord: metadata.DNShardRecord{
				ShardID: info.ShardID,
			},
			ReplicaID: info.ReplicaID,
			Address:   store.ServiceAddress,
		})
	}
```
since `store.Shards` is nil, this line will panic.
I don't think it is reasonable to prolong this wait time as 10 seconds is long enough. I think this error needs to be fixed in CN part,